types/logger/logger.go

```diff
@@ -42,3 +46,61 @@ func (w funcWriter) Write(p []byte) (int, error) {
 
 // Discard is a Logf that throws away the logs given to it.
 func Discard(string, ...interface{}) {}
+
+type limitData struct {
+	lim          *rate.Limiter
+	lastAccessed time.Time
+	msgBlocked   bool
+}
+
+// RateLimitedFn implements rate limiting by fstring on a given Logf.
+// Messages are allowed through at a maximum of r messages/second, in
+// bursts of up to b messages at a time.
+func RateLimitedFn(logf Logf, f float64, b int) Logf {
+	r := rate.Limit(f)
+	msgList := make(map[string]limitData)
+	lastPurge := time.Now()
+
+	rlLogf := func(s string, args ...interface{}) {
+		if rl, ok := msgList[s]; ok {
+
+			// Fields of structs contained in maps can't be modified; this is
+			// the workaround. See issue https://github.com/golang/go/issues/3117
+			temp := msgList[s]
```

Comment:
so far all loggers have been safe for concurrent use. This would be the first one that's not.

We should probably document that requirement more explicitly (on the `Logf` type above where it says it's `log.Printf`-like)

So you'd need to close over a mutex to protect the logger's state.